### PR TITLE
[QAOA (Secondary)] Adds Cost Hamiltonian Components

### PR DESCRIPTION
**Pull request by Lucaman99**
**Context:**

Part of the ongoing effort to add QAOA functionality to PennyLane

**Description of the Change:**

- Adds "cost Hamiltonian components", which are smaller "sub-Hamiltonians" that can be used to construct specific cost Hamiltonians, corresponding to different optimization problems.
- Updates the built-in `qaoa.maxcut` method to take advantage of these new "cost components".

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/qaoa/cost.py [hunk 1]
```diff
@@ -56,26 +208,15 @@ def maxcut(graph):
     >>> cost_h, mixer_h = qml.qaoa.maxcut(graph)
     >>> print(cost_h)
     >>> print(mixer_h)
-    (-0.5) [I0 I1] + (0.5) [Z0 Z1] + (-0.5) [I1 I2] + (0.5) [Z1 Z2]
+    (0.5) [Z0 Z1] + (0.5) [Z1 Z2] + (-1.0) [I0]
```
Comment (trbromley) [inline pennylane/qaoa/cost.py:211]: Minor thing: wouldn't it be like:
```pycon
>>> print
result
>>> print
result
```

File: pennylane/qaoa/cost.py [hunk 2]
```diff
@@ -56,26 +208,15 @@ def maxcut(graph):
...
-        obs.append(qml.PauliZ(node1) @ qml.PauliZ(node2))
-        coeffs.append(0.5)
+        raise ValueError("Input graph must be a nx.Graph, got {}".format(type(graph).__name__))
 
-    return (qml.Hamiltonian(coeffs, obs), qaoa.x_mixer(graph.nodes))
+    identity_h = qml.Hamiltonian(
+        [-0.5 for e in graph.edges], [qml.Identity(e[0]) @ qml.Identity(e[1]) for e in graph.edges]
+    )
+    H = edge_driver(graph, ["10", "01"]) + identity_h
+    return (H, qaoa.x_mixer(graph.nodes))
```
Comment (trbromley) [inline pennylane/qaoa/cost.py:222]: Was the approach before incorrect? It looks like more than just changing the method, but the actual result is different?

Comment (trbromley) [reply pennylane/qaoa/cost.py:222]: oh maybe just the identities were different :thinking:

**Review by trbromley — CHANGES_REQUESTED** [review]
Thanks @Lucaman99, great job with the code as always, just a couple suggestions on the docstrings.

File: pennylane/qaoa/cost.py (hunk 2 above)
Comment (Lucaman99) [reply pennylane/qaoa/cost.py:222]: @trbromley Yes, the only difference is the identities, which get added together (as the new approach to constructing this Hamiltonian involves addition of terms, which automatically simplifies the newly-created Hamiltonian)

**Comment by Lucaman99** [comment]
@trbromley @ixfoduap Hopefully the most recent changes make the docstring a bit clearer!

**Review by trbromley — APPROVED** [review]
Thanks @Lucaman99!

**Comment by josh146** [comment]
Likewise, I presume we're just waiting on @ixfoduap's review for this PR before it can be merged in?

**Comment by Lucaman99** [comment]
@josh146 Yes!

File: pennylane/qaoa/cost.py [hunk 3]
```diff
@@ -21,6 +21,174 @@
...
+
+    **Example**
+
+    As can be seen in the following example, ``"11"``, ``"10"``, and ``"01"`` will be assigned a lower
+    energy than ``"00"``:
+
+    >>> graph = nx.Graph([(0, 1), (1, 2)])
+    >>> hamiltonian = qaoa.edge_driver(graph, ["11", "10", "01"])
+    >>> print(hamiltonian)
+    (0.25) [Z0 Z1] + (0.25) [Z0] + (0.25) [Z1] + (0.25) [Z1 Z2] + (0.25) [Z2]
```
Comment (ixfoduap) [inline pennylane/qaoa/cost.py:95]: I'm still confused with the definition of this Hamiltonian, even after looking at the example.

1. What does it mean for an edge to be assigned an energy? _States_ (actually eigenstates of the Hamiltonian) can be associated with an energy, but the edges just denote a pair of qubits.

2. I don't think that the connection between the Hamiltonian and the reward is too clear. Maybe good to add in text below the code snippet what the expectation values are for some states? Something like
"For example, a quick calculation gives the following expectation values:
<000|H|000> = 1 (I think)
<110|H|110> = -0.25 (I think)
meaning, a colouring that assigns "00" to the first two qubits has higher energy than a colouring that assigns them with "11" "

File: pennylane/qaoa/cost.py [hunk 4]
```diff
@@ -21,6 +21,174 @@
...
+        using the :math:`|1\rangle` and :math:`|0\rangle` states of qubits. Therefore, given
+        some graph :math:`G`, each edge of the graph can be described by a pair of qubits, :math:`|00\rangle`,
+        :math:`|01\rangle`, :math:`|10\rangle`, or :math:`|11\rangle`, corresponding to the colourings of its endpoints.
+
+        When constructing QAOA cost functions, one must "penalize" certain states of the graph, and "reward"
+        others, by assigning higher and lower energies to these respective configurations. Given a set of vertex-colour
+        pairs (which each describe a possible  state of a graph edge), the ``edge_driver()``
+        function outputs a Hamiltonian that rewards the pairs in the set, and penalizes the others.
+
+        For example, given the set: :math:`\{|00\rangle, \ |01\rangle, \ |10\rangle\}` and the graph :math:`G`,
```
Comment (ixfoduap) [inline pennylane/qaoa/cost.py:112]:
```suggestion
        For example, given the reward set: :math:`\{|00\rangle, \ |01\rangle, \ |10\rangle\}` and the graph :math:`G`,
```

File: pennylane/qaoa/cost.py (hunk 3 above)
Comment (ixfoduap) [reply pennylane/qaoa/cost.py:95]: The usage details are helpful, but I think the explanation beforehand can be made clearer

File: pennylane/qaoa/cost.py [hunk 5]
```diff
@@ -21,6 +21,174 @@
...
+
+        where :math:`E(G)` is the set of edges of :math:`G`, and :math:`Z_i` is the Pauli-Z operator acting on the
+        :math:`i`-th wire. As can be checked, this Hamiltonian assigns an energy of :math:`-1/4` to the states
+        :math:`|00\rangle`, :math:`|01\rangle` and :math:`|10\rangle`, and an energy of :math:`3/4` to the state
+        :math:`|11\rangle`.
+
+        .. Note::
+
+            ``reward`` must always contain both :math:`|01\rangle` and :math:`|10\rangle`, or neither of the two.
+            Within a graph, there is no notion of "order"
```
Comment (ixfoduap) [inline pennylane/qaoa/cost.py:125]:
```suggestion
            Within an undirected graph, there is no notion of "order"
```

**Review by ixfoduap — CHANGES_REQUESTED** [review]
Looks solid! Left a few comments regarding clarity of the docstrings

File: pennylane/qaoa/cost.py (hunk 3 above)
Comment (Lucaman99) [reply pennylane/qaoa/cost.py]: Hopefully it's more clear now!

**Review by ixfoduap — APPROVED** [review]

**Pull request merged**